api/kv.go

```diff
@@ -0,0 +1,300 @@
+	}
+
+	return pathToRead, nil
+}
+
+func extractDataAndVersionMetadata(secret *Secret, version int) (*KVSecret, error) {
+	var data map[string]interface{}
+	var metadata *VersionMetadata
+	switch version {
+	case 1:
+		data = secret.Data
+		metadata = nil
+	case 2:
+		dataInterface, ok := secret.Data["data"]
+		if !ok {
+			return nil, fmt.Errorf("missing expected 'data' element")
+		}
+
+		if dataInterface == nil {
+			// this can happen when the secret has been deleted, but the metadata is still available
+			data = nil
+		} else {
+			data, ok = dataInterface.(map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("unexpected type for 'data' element: %T (%#v)", data, data)
+			}
+		}
+
+		metadataInterface, ok := secret.Data["metadata"]
+		if !ok {
+			return nil, fmt.Errorf("missing expected 'metadata' element")
+		}
+
+		metadataMap, ok := metadataInterface.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("unexpected type for 'metadata' element: %T (%#v)", metadata, metadata)
+		}
+
+		// deletion_time usually comes in as an empty string which can't be
+		// processed as time.RFC3339, so we reset it to a convertible value
+		if metadataMap["deletion_time"] == "" {
+			metadataMap["deletion_time"] = time.Time{}
+		}
+
+		d, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+			DecodeHook: mapstructure.StringToTimeHookFunc(time.RFC3339),
+			Result:     &metadata,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error setting up decoder for API response: %v", err)
+		}
+
+		err = d.Decode(metadataMap)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding metadata from API response into VersionMetadata: %v", err)
+		}
+	default:
+		return nil, fmt.Errorf("cannot parse secret without specifying valid KV secrets engine version")
+	}
+
+	return &KVSecret{
+		Data:     data,
+		Metadata: metadata,
+		Raw:      secret,
+	}, nil
+}
+
+func extractVersionMetadata(secret *Secret) (*VersionMetadata, error) {
+	metadataMap := secret.Data
+	var metadata *VersionMetadata
+
+	// deletion_time usually comes in as an empty string which can't be
+	// processed as time.RFC3339, so we reset it to a convertible value
+	if metadataMap["deletion_time"] == "" {
+		metadataMap["deletion_time"] = time.Time{}
+	}
+
+	d, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		DecodeHook: mapstructure.StringToTimeHookFunc(time.RFC3339),
+		Result:     &metadata,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error setting up decoder for API response: %v", err)
+	}
+
+	err = d.Decode(metadataMap)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding metadata from API response into VersionMetadata: %v", err)
+	}
+
+	return metadata, nil
+}
+
+func extractFullMetadata(secret *Secret) (*KVMetadata, error) {
+	var metadata *KVMetadata
+
+	// deletion_time usually comes in as an empty string which can't be
+	// processed as time.RFC3339, so we reset it to a convertible value
```

Comment:
If anyone knows how to get the DecodeHook to shut up and accept the empty string as time.RFC3339, lemme know. I tried the `,omitempty` and `,string` struct tags and still no good.